File: project/app/src/main/java/org/owntracks/android/ui/preferences/editor/EditorViewModel.java
package org.owntracks.android.ui.preferences.editor;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.Bindable;

import org.owntracks.android.BR;
import org.owntracks.android.R;
import org.owntracks.android.data.repos.WaypointsRepo;
import org.owntracks.android.injection.scopes.PerActivity;
import org.owntracks.android.model.messages.MessageConfiguration;
import org.owntracks.android.support.Parser;
import org.owntracks.android.support.Preferences;
import org.owntracks.android.ui.base.viewmodel.BaseViewModel;

import java.io.IOException;

import javax.inject.Inject;

import timber.log.Timber;


@PerActivity
public class EditorViewModel extends BaseViewModel<EditorMvvm.View> implements EditorMvvm.ViewModel<EditorMvvm.View> {
    private final Parser parser;
    private final Preferences preferences;
    @Bindable
    private String effectiveConfiguration;

    @Inject
    WaypointsRepo waypointsRepo;

    @Inject
    public EditorViewModel(Preferences preferences, Parser parser) {
        this.preferences = preferences;
        this.parser = parser; 
    }

    public void attachView(@Nullable Bundle savedInstanceState, @NonNull EditorMvvm.View view) {
        super.attachView(savedInstanceState, view);
        updateEffectiveConfiguration();
    }

    private void updateEffectiveConfiguration() {
        try {
            MessageConfiguration message = preferences.exportToMessage();
            message.setWaypoints(waypointsRepo.exportToMessage());
            message.set(preferences.getPreferenceKey(R.string.preferenceKeyPassword), "********");
            setEffectiveConfiguration(parser.toJsonPlainPretty(message));
        } catch (IOException e) {
            Timber.e(e);
            getView().displayLoadFailed();
        }
    }

    @Bindable
    public String getEffectiveConfiguration() {
        return effectiveConfiguration;
    }

    @Bindable
    private void setEffectiveConfiguration(String effectiveConfiguration) {
        this.effectiveConfiguration = effectiveConfiguration;
        notifyPropertyChanged(BR.effectiveConfiguration);
    }

    @Override
    public void onPreferencesValueForKeySetSuccessful() {
        updateEffectiveConfiguration();
        notifyPropertyChanged(BR.effectiveConfiguration);
    }
}

File: project/app/src/main/java/org/owntracks/android/ui/preferences/editor/EditorViewModel.kt
package org.owntracks.android.ui.preferences.editor

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.owntracks.android.data.waypoints.WaypointsRepo
import org.owntracks.android.model.Parser
import org.owntracks.android.preferences.Preferences
import org.owntracks.android.support.MessageWaypointCollection
import timber.log.Timber

@HiltViewModel
class EditorViewModel
@Inject
constructor(
    private val preferences: Preferences,
    private val parser: Parser,
    private val waypointsRepo: WaypointsRepo
) : ViewModel(), Preferences.OnPreferenceChangeListener {
  private val mutableEffectiveConfiguration = MutableStateFlow("")
  val effectiveConfiguration: StateFlow<String> = mutableEffectiveConfiguration

  private val mutableConfigLoadError = MutableStateFlow<Exception?>(null)
  val configLoadError: StateFlow<Exception?> = mutableConfigLoadError

  init {
    preferences.registerOnPreferenceChangedListener(this)
    updateEffectiveConfiguration()
  }

  override fun onCleared() {
    super.onCleared()
    preferences.unregisterOnPreferenceChangedListener(this)
  }

  val preferenceKeys = preferences.allConfigKeys.map { it.name }.toList()

  private fun updateEffectiveConfiguration() {
    viewModelScope.launch(Dispatchers.IO) {
      try {
        val message = preferences.exportToMessage()
        message.waypoints =
            MessageWaypointCollection().apply {
              addAll(waypointsRepo.getAll().map(waypointsRepo::fromDaoObject))
            }
        mutableEffectiveConfiguration.value = parser.toUnencryptedJsonPretty(message)
      } catch (e: Exception) {
        Timber.e(e)
        mutableConfigLoadError.value = e
      }
    }
  }

  fun setNewPreferenceValue(key: String, value: String) {
    preferences.importKeyValue(key, value)
  }

  override fun onPreferenceChanged(properties: Set<String>) {
    updateEffectiveConfiguration()
  }
}
